scrypto/src/component/stateful_access_rules.rs

```diff
@@ -0,0 +1,129 @@
+use radix_engine_derive::scrypto;
+use radix_engine_interface::api::api::SysNativeInvokable;
+use radix_engine_interface::api::types::{
+    AccessRulesOffset, ComponentId, GlobalAddress, RENodeId, SubstateOffset, ToString,
+};
+use radix_engine_interface::model::{
+    AccessRule, AccessRuleKey, AccessRuleSelector, AccessRules, AccessRulesSetAccessRuleInvocation,
+    AccessRulesSetMutabilityInvocation, ComponentAddress,
+};
+
+use crate::engine::scrypto_env::ScryptoEnv;
+use crate::runtime::{DataPointer, DataRef};
+
+use super::AccessRulesSubstate;
+
+// TODO: Should `Encode` and `Decode` be removed so that `StatefulAccessRules` can not be passed
+// between components?
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[scrypto(TypeId, Encode, Decode, Describe)]
+pub struct StatefulAccessRules {
+    component: ComponentIdentifier,
+    index: usize,
+}
+
+impl StatefulAccessRules {
+    pub(crate) fn new<T: Into<ComponentIdentifier>>(component: T, index: usize) -> Self {
+        Self {
+            component: component.into(),
+            index,
+        }
+    }
+
+    pub fn component_address(&self) -> ComponentIdentifier {
+        self.component.clone()
+    }
+
+    pub fn access_rules(&self) -> AccessRules {
```

Comment:
The problem with this is that AccessRules object won't reflect correct values if `set_method_auth` is called after `access_rules` call. If going down this route I'd rather expose a `RefMut` of AccessRules instead so that AccessRules may be mutated in place.

Reply:
Josh and I discussed this during a call. We agreed that in the long run it is desirable to use `RefMut` for cases like this. For the time being, we have agreed that Scrypto doesn't need to have the ability to read the full `AccessRules`. The only knowledge Scrypto needs is to know how many layers of `AccessRules` an entity has so that it can create `StatefulAccessRules` (name TBD) for each layer. 

For this, I implemented an `AccessRulesGetLengthInvocation` in [6d81f6a](https://github.com/radixdlt/radixdlt-scrypto/pull/624/commits/6d81f6aa6383f4443c18c87fa04b2a9bbe55cc7f)